### ISSUE TITLE
Unnecessary Comparison for Assertion

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#3437]
**Is your feature request related to a problem? Please describe.**
I had a recent bug that looked like,

```python
x = "foo"
y = ("foo",)

assert x == y
```

That assertion is always false as I got types mixed up. I turned on reportUnnecessaryComparison, but while
```python
if x == y:
  ...
```

is an error, `assert x == y` is not an error. For an assertion that's always false from comparison I'd think that's probably a bug and if a user really wants to fail on purpose they can assert False or similar.

**Describe the solution you'd like**
Always false comparisons in assertions should be reported as an error.

**Comment by erictraut** [comment]
The `reportUnnecessaryComparison` logic explicitly exempts `assert` statements. This is very intentional. When I initially implemented this support, I found that common uses of asserts triggered this diagnostic, producing a large number of false positives. Asserts are frequently used in test cases and to work around type issues (e.g. to apply narrowing).

As a quick experiment, I tried disabling the assert exemption and ran pyright on some internal Microsoft Python code bases, and (as I anticipated) it resulted in a massive number of false positives. 

So I don't think it makes sense to implement this as suggested.

**Comment by 5j9** [comment]
Can't Pyright distinguish between type-narrowing assertions and other invalid comparisons? I encountered an issue with code like the following, where it’s clear that no type narrowing is occurring:
```python
class MyCase:
    async def cache(self) -> str:
        return '0'

    async def run(self) -> None:
        assert self.cache() == '0'  # No warning, but will always raises AssertionError
        if self.cache() == '0':     # Warns as expected: reportUnnecessaryComparison
            ...
        ...
```
